src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs

```diff
@@ -37,24 +37,33 @@ public sealed partial class JsonDocument : IDisposable
         private JsonDocument(
             ReadOnlyMemory<byte> utf8Json,
             MetadataDb parsedData,
-            byte[]? extraRentedArrayPoolBytes = null,
-            PooledByteBufferWriter? extraPooledByteBufferWriter = null,
-            bool isDisposable = true)
+            bool isDisposable)
         {
             Debug.Assert(!utf8Json.IsEmpty);
 
-            // Both rented values better be null if we're not disposable.
-            Debug.Assert(isDisposable ||
-                (extraRentedArrayPoolBytes == null && extraPooledByteBufferWriter == null));
-
-            // Both rented values can't be specified.
-            Debug.Assert(extraRentedArrayPoolBytes == null || extraPooledByteBufferWriter == null);
-
             _utf8Json = utf8Json;
             _parsedData = parsedData;
+            IsDisposable = isDisposable;
+        }
+
+        private void TakePooledArrayOwnership(byte[] extraRentedArrayPoolBytes)
```

Comment:
I take it this is related to my earlier feedback about disposing the `JsonDocument` if validation fails? Why couldn't we just do that here? Initializing the object post-construction feels a bit error prone to me.

Reply:
It was causing double Return to the array pool. The caller who rents the array will return the array if parse fails with an exception but we also need to dispose the document when we throw from validate (since the metadata db could be rented) so the array gets returned twice.

If we want to avoid the split initialization we could remove the array return from the caller side and assume the core parse method gets ownership of the array when it gets called.

Reply:
I changed it to dispose the metadata database directly instead of disposing the document and reverted all the split initialization changes.